request.go

```diff
@@ -230,13 +219,20 @@ func fileget(h FileReader, r *Request, pkt requestPacket) responsePacket {
 // wrap FileWriter handler
 func fileput(h FileWriter, r *Request, pkt requestPacket) responsePacket {
 	var err error
-	writer := r.getWriter()
+	r.state.RLock()
+	writer := r.state.writerAt
+	r.state.RUnlock()
 	if writer == nil {
-		writer, err = h.Filewrite(r)
-		if err != nil {
-			return statusFromError(pkt, err)
+		r.state.Lock()
+		if r.state.writerAt == nil {
+			r.state.writerAt, err = h.Filewrite(r)
+			if err != nil {
+				r.state.Unlock()
```

Comment:
I had recommended on the other PR moving this code into the `getReader` / `getWriter` function to be able to use the `defer r.state.Unlock()` so that one can avoid having to ensure that one puts unlocks in every code path.

It’s not that this code is wrong, it’s rather just a defensive coding tactic that I like to follow.

Reply:
Normally I agree, but I wanted to group the code together like that as I thought it'd help make the refactoring simpler by eliminating the extra indirection. I also noticed in my testing that the defer still adds about a 50% overhead to the calls and thought I'd go with the faster method until I had a chance to test how it might affect performance/throughput more thoroughly.

Reply:
:+1: sounds good.